### ISSUE TITLE
overscroll code doesn't work properly anymore

### ISSUE THREAD

**Issue by paulrouget** [servo/servo#19718]
/cc @glennw 

If window::window_rect() returns `TypedRect::new(TypedPoint2D::new(0, 50), size)`, the page size is correct, but its position is wrong (not shifted down to 50px).

Here is a screenshot:

<img width="890" alt="screen shot 2018-01-08 at 04 31 29" src="https://user-images.githubusercontent.com/373579/34658156-d2952d1a-f42c-11e7-94fb-b252af8a7db2.png">

The whole page should be shifted down.

Could this be the consequence of what @mrobinson was mentioning here: https://github.com/servo/webrender/issues/2159#issuecomment-349925639

**Comment by paulrouget** [comment]
STR:

Change this in /ports/:

```diff
diff --git a/ports/glutin/window.rs b/ports/glutin/window.rs
index f2f16b0600..e2dbe533c3 100644
--- a/ports/glutin/window.rs
+++ b/ports/glutin/window.rs
@@ -996,8 +996,10 @@ impl WindowMethods for Window {
     }

     fn window_rect(&self) -> DeviceUintRect {
-        let size = self.framebuffer_size();
-        let origin = TypedPoint2D::zero();
+        let mut size = self.framebuffer_size();
+        size.height -= 50;
+
+        let origin = TypedPoint2D::new(0, 50);
         DeviceUintRect::new(origin, size)
     }
```

Load this page:

```
<style>
  * {
    margin: 0;
    padding: 0;
  }
  div, p {
    position: absolute;
    top: 0;
    left: 0;
  }
  body { background: yellow; }
  div {
    width: 100%;
    height: 100%;
    background: red;
  }
  p {
    width: 100px;
    height: 100px;
    background: green;
  }
</style>

<div></div>
<p></p>
```

Good:

<img width="483" alt="screen shot 2018-01-08 at 09 52 19" src="https://user-images.githubusercontent.com/373579/34663937-a57221e4-f459-11e7-91d1-72f7b2a37b02.png">

Bad:

<img width="368" alt="screen shot 2018-01-08 at 09 44 38" src="https://user-images.githubusercontent.com/373579/34663944-a87dc2bc-f459-11e7-8979-807d8a5baf0f.png">

**Comment by paulrouget** [comment]
Bisecting tells me the https://github.com/servo/servo/commit/cdc163533c23e40a5faef75088fc7a0381f6dc75 is the issue.

**Comment by paulrouget** [comment]
@mrobinson / @glennw does that help or do you also need me to bisect the webrender changes?

**Comment by paulrouget** [comment]
Let me describe a bit more what the expected behavior is.

I'm not sure why we call that `overscroll` as it's not directly related to scrolling. We want to be able to draw outside of the DOM window:

The HTML page is usually clipped by the DOM window dimension (x:0;y:0 is the CSS top left corner of the page, but also where the page is clipped). But we might want to draw the HTML content outside of the DOM window. A typical example is the ability to see the background of the page below a toolbar. Here is Gif that shows the effect:

https://cloud.githubusercontent.com/assets/373579/23516132/9f437e2c-ff6c-11e6-8340-97c7de94b411.gif

(the page is painted below the toolbar as weel)

So we want to be able to clip an area that is larger than the DOM window.

To do so, in https://github.com/servo/webrender/pull/951, we introduced the notion of `inner_rect` which represent the DOM window, and `window_size` the represent the region to actually clip.

What is important is to not mess up the layout. In the case of the toolbar example, we want a `position:fixed;top:0` element to be painted under (in `y` coordinates) the toolbar, while the top scrolled-out content would be painted below (in `z` coordinates) the toolbar.

I hope this clarify the behavior.

Do not hesitate to ask for clarifications.

**Comment by mrobinson** [comment]
@paulrouget I'm pretty sure I have tracked down the PR that caused this (servo/webrender#2043). Is this issue still relevant now that work on BrowserHTML is ceasing? If it is, I can try to fix it in the upcoming weeks.

**Comment by paulrouget** [comment]
>  Is this issue still relevant now that work on BrowserHTML is ceasing? 

Yes. BrowserHTML doesn't use this feature. This is used by regular embedders.

> I can try to fix it in the upcoming weeks.

It is not a high priority, but that would be great to get that fixed soon.

Thank you!

**Comment by mrobinson** [comment]
Overscroll / rubberband scrolling doesn't exist any longer. We'll likely need to re-implement it, but for now we can close this issue.
